wordpress/mod.ts

```diff
@@ -36,174 +127,137 @@ export default function (userOptions?: Partial<Options>) {
+    this.#limit = options.limit;
+    this.#maxPerPage = options.maxPerPage;
+    this.#collections = options.collections;
+    this.#customTransformer = options.transform;
 
-  /** Returns all tags of the site */
-  async *tags(limit?: number): AsyncGenerator<Partial<PageData>> {
-    for await (const tag of this.#fetchAll<WP_REST_API_Term>("tags", limit)) {
-      yield termToData(tag);
+    if (options.auth) {
+      this.#pass = options.auth.includes(":")
+        ? btoa(options.auth)
+        : options.auth;
     }
   }
 
-  /** Returns all authors of the site */
-  async *authors(limit?: number): AsyncGenerator<Partial<PageData>> {
-    for await (const user of this.#fetchAll<WP_REST_API_User>("users", limit)) {
-      yield userToData(user);
-    }
-  }
+  async *collection(
+    name: string,
+    limit: number = this.#limit,
+  ): AsyncGenerator<Partial<PageData>> {
+    const filterList: Transformer[] = [
+      requirementFilter,
+      presetTransform[name] || this.#customTransformer["*"],
+      this.#customTransformer[name],
+    ];
 
-  /** Returns all media of the site */
-  async *media(limit?: number): AsyncGenerator<Partial<PageData>> {
-    for await (
-      const media of this.#fetchAll<WP_REST_API_Attachment>("media", limit)
-    ) {
-      yield mediaToData(media);
+    const response = this.#fetchAll<WP_REST_API>(
+      this.#collections[name],
+      limit,
+    );
+    for await (const raw of response) {
+      yield filterList.reduce(
+        (data, transform) =>
+          typeof transform === "function" ? transform(data, raw) : data,
+        { type: name } as Partial<PageData>,
+      );
     }
   }
 
-  async *#fetchAll<T>(path: string, limit = this.#limit): AsyncGenerator<T> {
-    const max_per_page = 100;
-    let page = 1;
+  async *#fetchAll<T>(
+    apiPath: string,
+    limit: number,
+  ): AsyncGenerator<T> {
+    let pageIndex = 1;
     let counter = 0;
 
     while (limit > counter) {
-      const per_page = Math.min(limit - counter, max_per_page);
-      counter += per_page;
+      const pageSize = Math.min(limit - counter, this.#maxPerPage);
+      counter += pageSize;
+
       const posts = await this.#fetch<T>(
-        `${path}?per_page=${per_page}&page=${page++}`,
+        apiPath,
+        [
+          ["page", `${pageIndex++}`],
+          ["per_page", `${pageSize}`],
+        ],
       );
 
       for (const post of posts) {
         yield post;
       }
 
-      if (posts.length < max_per_page) {
+      if (posts.length < this.#maxPerPage) {
         break;
       }
     }
   }
 
-  async #fetch<T>(path: string): Promise<T[]> {
-    const url = new URL(path, this.#api_endpoint);
+  async #fetch<T>(apiPath: string, query: QueryParams = []): Promise<T[]> {
+    const url = this.#craftUrl(apiPath, query);
     const cacheKey = url.toString();
 
     if (this.#cache.has(cacheKey)) {
-      return this.#cache.get(cacheKey);
+      return this.#cache.get(cacheKey) as T[];
```

Comment:
Not sure this cache is useful, because it lives on memory, and it's not likely to request the same URL twice (once the page is generated, why we would like to generate it again?).

I think a persistent memory would be better. Lume has [the `read` helper ](https://github.com/lumeland/lume/blob/master/core/utils.ts#L486) that uses the Web Cache API to cache the HTTP responses. The only problem with this API is it's not possible to define a TTL, so the response is cached indefinitely.

One possible solution I've implemented successfuly [in this other package](https://github.com/oom-components/mastodon-comments/blob/main/src/comments.js#L120-L121) is to store in a header the time when the response has stored, so we can invalidate the cache at any point.

This would allow to run the plugin even offline, using the cached responses as fallback.

Reply:
I'm not very good at caching things, so I will go with your suggestion. It would be great if we have any cache system that can work well with Cloudflare Page deployment.

Reply:
Great. I'm thinking of including this function in the utils of lume, so it could be used in this way:

```ts
import { fetch } from "lume/core/utils.ts";

const response = await fetch(url, 3600); // cache 1 hour (3600 seconds).
```

Reply:
Please don't forget the fetch Headers because the auth feature depends on it :D